src/components/vr_module/include/vr_module/event_engine/event_dispatcher.h

```diff
@@ -0,0 +1,206 @@
+ without specific prior written permission.
+
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef  SRC_COMPONENTS_VR_MODULE_INCLUDE_VR_MODULE_EVENT_ENGINE_EVENT_DISPATCHER_H_
+#define  SRC_COMPONENTS_VR_MODULE_INCLUDE_VR_MODULE_EVENT_ENGINE_EVENT_DISPATCHER_H_
+
+#include <list>
+#include <map>
+
+#include "utils/lock.h"
+#include "utils/singleton.h"
+
+#include "vr_module/event_engine/event.h"
+#include "vr_module/event_engine/event_observer.h"
+
+#include "interfaces/MOBILE_API.h"
+
+namespace event_engine {
+
+template<typename EventMessage, typename EventID>
+class EventDispatcher :
+    public utils::Singleton<EventDispatcher<EventMessage, EventID>> {
+ public:
+  /**
+   * @brief Delivers the event to all subscribers
+   *
+   * @param event Received event
+   */
+  void raise_event(const Event<EventMessage, EventID>& event);
+
+  /**
+   * @brief Subscribe the observer to event
+   *
+   * @param event_id    The event ID to subscribe for
+   * @param correlation_id  The event correlation ID
+   * @param observer    The observer to subscribe for event
+   */
+  void add_observer(const EventID& event_id,
+                    int32_t correlation_id,
+                    EventObserver<EventMessage, EventID>* const observer);
+
+  /**
+   * @brief Unsubscribes the observer from specific event
+   *
+   * @param event_id    The event ID to unsubscribe from
+   * @param observer    The observer to be unsubscribed
+   */
+  void remove_observer(const EventID& event_id,
+                    const EventObserver<EventMessage, EventID>* const observer);
+
+  /**
+   * @brief Unsubscribes the observer from all events
+   *
+   * @param observer  The observer to be unsubscribed
+   */
+  void remove_observer(
+      const EventObserver<EventMessage, EventID>* const observer);
+
+ private:
+  /**
+   * @brief Default constructor
+   */
+  EventDispatcher();
+
+  /**
+   * @brief Destructor
+   */
+  virtual ~EventDispatcher();
+
+
+  FRIEND_BASE_SINGLETON_CLASS(EventDispatcher);
+
+  // Data types section
+  typedef std::list<EventObserver<EventMessage, EventID>*> ObserverList;
+  typedef std::map<int32_t, ObserverList>      ObserversMap;
+  typedef std::map<EventID, ObserversMap>      EventObserverMap;
```

Comment:
Typedef should be the first according with [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Declaration_Order)